qmi/instruments/bristol/fos.py

```diff
@@ -1,98 +1,248 @@
+
+        Parameters:
+            unique_id:         A unique ID string.
+
+        Returns:
+            device_descriptor: The device descriptor with matching unique ID or None.
+        """
+        assert ul is not None and enums is not None
+        device_descriptors = ul.get_daq_device_inventory(enums.InterfaceType.ANY)
+        for device_descriptor in device_descriptors:
+            print(device_descriptor.unique_id)
+            if device_descriptor.unique_id == unique_id:
+                return device_descriptor
+
+        return None  # Device not found.
+
+    @rpc_method
+    def open(self) -> None:
+        assert ul is not None and enums is not None
+        ul.ignore_instacal()  # With this we ignore 'cg.cfg' file and enable runtime configuring.
+        device_descriptor = self._find_device_descriptor(self._unique_id)
+        if device_descriptor is None:
+            _logger.error("Bristol FOS with unique_id '%s' not found.", self._unique_id)
+            raise ValueError(f"Bristol FOS with unique_id '{self._unique_id!r}' not found.")
+
+        try:
+            ul.create_daq_device(self.board_id, device_descriptor)
+            assert self.board_id == ul.get_board_number(
+                device_descriptor
+            ), f"{self.board_id} != {ul.get_board_number(device_descriptor)}"
+            ul.d_config_port(self.board_id, enums.DigitalPortType.FIRSTPORTA, enums.DigitalIODirection.OUT)
+
+        except Exception as exc:
+            _logger.error("Bristol FOS device configuration failed with: %s", str(exc))
+            ul.release_daq_device(self.board_id)
+            raise qmi.core.exceptions.QMI_InstrumentException("Bristol FOS device port configuration failed.") from exc
+
+    @rpc_method
+    def close(self) -> None:
+        assert ul is not None
+        ul.release_daq_device(self.board_id)
+
+    @rpc_method
+    def select_channel(self, channel: int) -> None:
+        assert ul is not None
+        # Note that the 'channel parameter has values 1..4 ; these are mapped to value 0..3 here.
+        ul.d_out(self._board_id, enums.DigitalPortType.FIRSTPORTA, channel - 1)
+
+
+class Bristol_Fos(QMI_Instrument):
+    """Base class for the Bristol FOS instrument driver"""
+
+    def __init__(self, context: QMI_Context, name: str, unique_id: str, board_id: int = 0) -> None:
+        """Base class initialization. Depending on the system platform, a Windows-compatible or
+        Unix-compatible driver version is instantiated.
+
+        Attributes:
+            fos:        The instantiated FOS device on the driver.
+
+        Parameters:
+            context:    A QMI_Context instance.
+            name:       Name for the instrument in the context.
+            unique_id:  An unique identification number, a.k.a. serial number of the device.
+            board_id:   Board number for Windows driver. Not used for Unix driver.
```

Comment:
if `board_id` is not used for unix, should we make it optional instead of setting a default value of 0. Because 0 does hold meaning when its a Windows FOS, so if not set then a value of 0 is still valid.

Reply:
I think it is simpler this way. We simply do not forward the `board_id` to the Unix driver version, and then we don't need to do any `assert is not None` step before forwarding it to the Windows version.